vnext/include/ReactUWP/Views/ShadowNodeBase.h

```diff
@@ -122,6 +137,24 @@ struct REACTWINDOWS_EXPORT ShadowNodeBase : public facebook::react::ShadowNode
   bool m_onMouseEnter = false;
   bool m_onMouseLeave = false;
   bool m_onMouseMove = false;
+
+  // Support Keyboard
+public:
+  void UpdateHandledKeyboardEvents(KeyboardType forKeyDown, folly::dynamic const& value);
+
+private:
+  bool ShouldHookKeyboardEvent(XamlView const& xamlView);
+  void UpdateKeyboardEventHooks(XamlView const& xamlView);
+
+  void KeyboardEventHandledHandler(KeyboardEventPhase phase, winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const& args);
+  bool ShouldMarkKeyboardHandled(std::vector<KeyboardEvent> const& handledEvents, KeyboardEvent currentEvent);
+  std::vector<KeyboardEvent> m_handledKeyUpKeyboardEvents;
+  std::vector<KeyboardEvent> m_handledKeyDownKeyboardEvents;
```

Comment:
these vectors are also an additional 2 heap allocs constructing every shadownode.  Also not sure why these aren't members of the KeyboardEventHandler

Reply:
Good call

---
In reply to: [295400144](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295400144) [](ancestors = 295400144)

Reply:
I moved them out of ShadowNodeBase.